ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -63,44 +61,142 @@ impl Distribution<f64> for RoundedBoxMuller {
+        if probability < f64::MIN_POSITIVE {
+            return Err(Error::BadGeometricProb(probability));
+        }
+        Ok(Self { probability })
+    }
+    /// Generates a sample from the `Geometric` distribution.
+    pub fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        // Create a Bernoulli distribution with the specified success probability
+        let bernoulli = Bernoulli::new(self.probability).expect("Invalid probability");
+        // Generate Bernoulli random numbers until the first success
+        let mut attempts = 0;
+        while !bernoulli.sample(rng) {
+            attempts += 1;
+        }
+        attempts
+    }
+}
+impl Distribution<u32> for Geometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        self.sample(rng)
+    }
+}
+
+/// Double Geometric
+#[derive(Debug, PartialEq)]
+pub struct DoubleGeometric {
+    shift: u32,
+    geometric: Geometric,
+}
+impl DoubleGeometric {
+    /// Creates a new `DoubleGeometric` distribution with the given success probability and shift parameter.
+    pub fn new(s: f64, shift: u32) -> Result<Self, Error> {
+        if s < f64::MIN_POSITIVE {
+            return Err(Error::BadS(s));
+        }
+        let success_probability = 1.0 - E.powf(-1.0 / s);
+        Ok(Self {
+            shift,
+            geometric: Geometric::new(success_probability)?,
+        })
+    }
+    /// Generates a sample from the `DoubleGeometric` distribution
+    pub fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i32 {
+        // Self::generate_double_geometric(self.s, self.shift, rng)
+        let attempts1 = self.geometric.sample(rng);
+        let attempts2 = self.geometric.sample(rng);
+        (self.shift + attempts1 - attempts2).try_into().unwrap()
+    }
+}
+impl Distribution<i32> for DoubleGeometric {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> i32 {
+        self.sample(rng)
+    }
+}
+
+/// Truncated Double Geometric distribution.
+#[derive(Debug, PartialEq)]
+pub struct TruncatedDoubleGeometric {
+    shift: u32, // is truncated to [0, 2*shift]
+    double_geometric: DoubleGeometric,
+}
+impl TruncatedDoubleGeometric {
+    /// Creates a new `TruncatedDoubleGeometric` distribution with the given success probability and shift parameter.
+    pub fn new(s: f64, shift: u32) -> Result<Self, Error> {
+        if s < f64::MIN_POSITIVE {
+            return Err(Error::BadS(s));
+        }
+        Ok(Self {
+            shift,
+            double_geometric: DoubleGeometric::new(s, shift)?,
+        })
+    }
+    /// Generates a sample from the `TruncatedDoubleGeometric` distribution.
+    pub fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u32 {
+        // generate_truncated_double_geometric(self.s, self.shift, rng)
+        let mut reject = 1;
+        let mut sample = 0; // Declare sample here
+        while reject == 1 {
+            sample = self.double_geometric.sample(rng); // Assign a value to sample inside the loop
```

Comment:
this comment is moot as code is self explanatory. It is recommended to avoid such comments because they distract readers